### PR TITLE
ramips: add support for Netgear R6350

### PR DESCRIPTION
**Pull request by drvlabo**
Netgear R6350 is a wireless router, aka Netgear AC1750.

Specification:
- SoC: Mediatek MT7621AT (2 CPU cores, 4 threads)
- RAM: 128MiB (Nanya NT5CC64M16GP-DI)
- ROM: 128MiB NAND Flash (Macronix MX30LF1G18AC-TI)
- Wireless:
   for 11b/g/n (upto 300Mbps): MT7603
   for 11a/ac  (upto 1450Mbps) : MT7615, is not avaliable now
- Ethernet LAN speed: up to 1000Mbps
- Ethernet LAN ports: 4
- Ethernet WAN speed: up to 1000Mbps
- Ethernet WAN ports: 1
- USB ports: 1 (USB 2.0)
- LEDs: 4 (all can be controlled by SoC's GPIO)
- buttons: 2
- serial ports: unknown

Installation through telnet:
  - Copy kernel.bin and rootfs.bin to a USB flash disk,
    plug to usb port on the router.
  - Enable telnet with link: http://192.168.1.1/setup.cgi?todo=debug
    (login if required, default: admin password)
  - You will see "Debug Enabled!"
  - Telnet 192.168.1.1 and login with "root"
  - ls /mnt/shares/ to find out path of your USB disk.
    'myUdisk' for example.
  - cd /mnt/shares/myUdisk
  - mtd_write write rootfs.bin Rootfs
  - mtd_write write kernel.bin Kernel
  - reboot

recovery when bricked:
  nmrpflash can be used to recover to the netgear firmware
  if a broken image was flashed.

Signed-off-by: NOGUCHI Hiroshi <drvlabo@gmail.com>

### CODE REVIEW

**Comment by mkresin** [comment]
I'm really concerted about your partitions. At least since #1304 we known that some netgear devices are using an on flash partition map. And "SC PART_MAP" pretty much looks like this on flash partition map.

Please grep the tree for `ecoscentric,redboot-fis-partitions` and `brcm,trx` to find some example code and use of such on flash partition maps.

**Comment by drvlabo** [comment]
@mkresin Thanks for comment.
I'll investigate the data structure of Netgear's partition map.

**Comment by drvlabo** [comment]
@mkresin I implemented partition parser/builder for "SC_PART_MAP".
`[    3.339227] scpart: signature found at 0x100000`
`[    3.375159] scpart: valid 'SC PART MAP' found (29 partitions)`
`[    3.386639] 5 scpart partitions found on MTD device MT7621-NAND`
`[    3.398428] Creating 5 MTD partitions on "MT7621-NAND":`
`[    3.408848] 0x000000000000-0x000000100000 : "u-boot"`
`[    3.419746] 0x000000100000-0x000000200000 : "SC PART_MAP"`
`[    3.431391] 0x000000200000-0x000000600000 : "kernel`"
`[    3.442147] 0x000000600000-0x000002e00000 : "ubi"`
`[    3.452650] 0x000004620000-0x000004820000 : "factory"`

But, in fact, "factory" partition lies 0x04600000 - 0x04800000 .
In other words, partition definitions in "SC_PART_MAP" does not match the actual FLASH usage in target.
Therefore, I think that we must write the partition map manually to use "fixed-partitions" parser.

**Comment by mkresin** [comment]
> In other words, partition definitions in "SC_PART_MAP" does not match the actual FLASH usage in target.

I'm quite sure you will agree to a "what the hell". I've really no idea what's the deal with this board.

Would you please add a bootlog of the original firmware for reference. It might be a good idea to paste your existing code into a comment (for later use). Maybe it is required for another board.

>  Therefore, I think that we must write the partition map manually to use "fixed-partitions" parser.

As long as no one else has an explanation what is going on here, I agree that using fixed-partitions is the way to go.

**Comment by drvlabo** [comment]
> Would you please add a bootlog of the original firmware for reference. 
Unfortunately I failed to attach serial console, so I can not obtain a bootlog. Sorry ...

Current implemented codes are here:
```
diff --git a/target/linux/ramips/dts/R6350.dts b/target/linux/ramips/dts/R6350.dts
index ea3eeda747..1270ce15f9 100644
--- a/target/linux/ramips/dts/R6350.dts
+++ b/target/linux/ramips/dts/R6350.dts
@@ -79,7 +79,7 @@
 	status = "okay";
 
 	partitions {
-		compatible = "fixed-partitions";
+		compatible = "sercomm,sc-partitions", "fixed-partitions";
 		#address-cells = <1>;
 		#size-cells = <1>;
 
@@ -87,22 +87,26 @@
 			label = "u-boot";
 			reg = <0x0 0x100000>;
 			read-only;
+			scpart-id = <0>;
 		};
 
 		partition@100000 {
 			label = "SC PART_MAP";
 			reg = <0x100000 0x100000>;
 			read-only;
+			scpart-id = <1>;
 		};
 
 		partition@200000 {
 			label = "kernel";
 			reg = <0x200000 0x400000>;
+			scpart-id = <2>;
 		};
 
 		partition@600000 {
 			label = "ubi";
 			reg = <0x600000 0x2800000>;
+			scpart-id = <3>;
 		};
 
 		/* +0x02E00000 - +0x04600000 : reserved */
@@ -111,6 +115,7 @@
 			label = "factory";
 			reg = <0x4600000 0x200000>;
 			read-only;
+			scpart-id = <16>;
 		};
 
 		/* +0x04800000 - +0x08000000 : reserved */
diff --git a/target/linux/ramips/mt7621/config-4.14 b/target/linux/ramips/mt7621/config-4.14
index 0e06023780..7b53334ca4 100644
--- a/target/linux/ramips/mt7621/config-4.14
+++ b/target/linux/ramips/mt7621/config-4.14
@@ -166,6 +166,7 @@ CONFIG_MODULES_USE_ELF_REL=y
 CONFIG_MT7621_WDT=y
 # CONFIG_MTD_CFI_INTELEXT is not set
 CONFIG_MTD_CMDLINE_PARTS=y
+CONFIG_MTD_SERCOMM_PARTS=y
 CONFIG_MTD_M25P80=y
 CONFIG_MTD_NAND=y
 CONFIG_MTD_NAND_ECC=y
diff --git a/target/linux/ramips/patches-4.14/0055-mtd-scpart-parser.patch b/target/linux/ramips/patches-4.14/0055-mtd-scpart-parser.patch
new file mode 100644
index 0000000000..a1d2cbfee7
--- /dev/null
+++ b/target/linux/ramips/patches-4.14/0055-mtd-scpart-parser.patch
@@ -0,0 +1,289 @@
+diff -Nurp -x '*.o' -x '.*.cmd' linux41467_a/drivers/mtd/Kconfig linux41467_b/drivers/mtd/Kconfig
+--- linux41467_a/drivers/mtd/Kconfig	2018-09-02 14:56:16.154460632 +0900
++++ linux41467_b/drivers/mtd/Kconfig	2018-09-02 20:57:19.410600169 +0900
+@@ -174,6 +174,12 @@ config MTD_BCM47XX_PARTS
+ 	  This provides partitions parser for devices based on BCM47xx
+ 	  boards.
+ 
++config MTD_SERCOMM_PARTS
++	tristate "SERCOMM partitioning information support"
++	depends on MTD_OF_PARTS
++	help
++	  This provides partition table parser for SERCOMM partition map
++
+ menu "Partition parsers"
+ source "drivers/mtd/parsers/Kconfig"
+ endmenu
+diff -Nurp -x '*.o' -x '.*.cmd' linux41467_a/drivers/mtd/Makefile linux41467_b/drivers/mtd/Makefile
+--- linux41467_a/drivers/mtd/Makefile	2018-09-02 14:56:16.155460624 +0900
++++ linux41467_b/drivers/mtd/Makefile	2018-09-02 20:54:49.656754635 +0900
+@@ -17,6 +17,7 @@ obj-$(CONFIG_MTD_AR7_PARTS)	+= ar7part.o
+ obj-$(CONFIG_MTD_BCM63XX_PARTS)	+= bcm63xxpart.o
+ obj-$(CONFIG_MTD_BCM47XX_PARTS)	+= bcm47xxpart.o
+ obj-$(CONFIG_MTD_MYLOADER_PARTS) += myloader.o
++obj-$(CONFIG_MTD_SERCOMM_PARTS) += scpart.o
+ obj-y				+= parsers/
+ 
+ # 'Users' - code which presents functionality to userspace.
+diff -Nurp -x '*.o' -x '.*.cmd' linux41467_a/drivers/mtd/scpart.c linux41467_b/drivers/mtd/scpart.c
+--- linux41467_a/drivers/mtd/scpart.c	1970-01-01 09:00:00.000000000 +0900
++++ linux41467_b/drivers/mtd/scpart.c	2018-09-07 12:54:02.086173491 +0900
+@@ -0,0 +1,258 @@
++/*
++ *
++ * This program is free software; you can redistribute it and/or modify
++ * it under the terms of the GNU General Public License as published by
++ * the Free Software Foundation; either version 2 of the License, or
++ * (at your option) any later version.
++ *
++ * This program is distributed in the hope that it will be useful,
++ * but WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
++ * GNU General Public License for more details.
++ *
++ * You should have received a copy of the GNU General Public License
++ * along with this program; if not, write to the Free Software
++ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
++ *
++ * Sercomm/Netgear FLASH partition table.
++ */
++
++#define	DEBUG
++
++#include <linux/kernel.h>
++#include <linux/slab.h>
++#include <linux/mtd/mtd.h>
++#include <linux/mtd/partitions.h>
++#include <linux/module.h>
++
++
++#define	MOD_NAME	"scpart"
++
++static const char sc_part_magic[] = {
++	'S', 'C', 'F', 'L', 'M', 'A', 'P', 'O', 'K', '\0',
++};
++#define	PART_MAGIC_LEN	sizeof(sc_part_magic)
++
++/* assumes that all fields are set by CPU native endian */
++struct sc_part_desc {
++	uint32_t	part_id;
++	uint32_t	part_offs;
++	uint32_t	part_bytes;
++};
++#define	ID_ALREADY_FOUND	(0xFFFFFFFFUL)
++
++#define	MAP_OFFS_IN_BLK	(0x800)
++
++#define	MAP_MIRROR_NUM	(2)
++
++static int scpart_desc_is_valid(struct sc_part_desc *pdesc)
++{
++	return !!((pdesc->part_id != 0xFFFFFFFFUL) &&
++			(pdesc->part_offs != 0xFFFFFFFFUL) &&
++			(pdesc->part_bytes != 0xFFFFFFFFUL));
++}
++
++static int scpart_scan_partmap(struct mtd_info *master, loff_t partmap_offs,
++				struct sc_part_desc **ppdesc)
++{
++	uint8_t *buf;
++	loff_t offs;
++	size_t retlen;
++	struct sc_part_desc *pdesc = NULL;
++	struct sc_part_desc *tmpdesc;
++	int cnt = 0;
++	int res2;
++	int res = 0;
++
++	buf = kzalloc(master->erasesize, GFP_KERNEL);
++	if (!buf) {
++		res = -ENOMEM;
++		goto out;
++	}
++
++	res2 = mtd_read(master, partmap_offs, master->erasesize, &retlen, buf);
++	if (res2 || (retlen != master->erasesize)) {
++		res = -EIO;
++		goto free;
++	}
++
++	offs = MAP_OFFS_IN_BLK;
++	while((offs + sizeof(*tmpdesc)) < master->erasesize) {
++		tmpdesc = (struct sc_part_desc*)&(buf[offs]);
++		if (!scpart_desc_is_valid(tmpdesc))
++			break;
++		cnt++;
++		offs += sizeof(*tmpdesc);
++	}
++
++	if (cnt > 0) {
++		int bytes = cnt * sizeof(*pdesc);
++
++		pdesc = kzalloc(bytes, GFP_KERNEL);
++		if (!pdesc) {
++			res = -ENOMEM;
++			goto free;
++		}
++		memcpy(pdesc, &(buf[MAP_OFFS_IN_BLK]), bytes);
++
++		*ppdesc = pdesc;
++		res = cnt;
++	}
++
++  free:
++	kfree(buf);
++
++  out:
++	return res;
++}
++
++static int scpart_find_partmap(struct mtd_info *master,
++				struct sc_part_desc **ppdesc)
++{
++	loff_t offs;
++	uint8_t rdbuf[PART_MAGIC_LEN];
++	size_t retlen;
++	int magic_found = 0;
++	int res2;
++	int res = 0;
++
++	offs = 0;
++	while((magic_found < MAP_MIRROR_NUM) &&
++			(offs < master->size) && !mtd_block_isbad(master, offs)) {
++		res2 = mtd_read(master, offs, PART_MAGIC_LEN, &retlen, rdbuf);
++		if (res2 || (retlen != PART_MAGIC_LEN)) {
++			res = -EIO;
++			goto out;
++		}
++		if (!memcmp(rdbuf, sc_part_magic, PART_MAGIC_LEN)) {
++			pr_debug("%s: signature found at 0x%llx\n", MOD_NAME, offs);
++			magic_found++;
++			res = scpart_scan_partmap(master, offs, ppdesc);
++			if (res > 0)
++				goto out;
++		}
++		offs += master->erasesize;
++	}
++
++  out:
++	if (res > 0)
++		pr_info("%s: valid 'SC PART MAP' found (%d partitions)\n", MOD_NAME, res);
++	else
++		pr_info("%s: no valid 'SC PART MAP'\n", MOD_NAME);
++
++	return res;
++}
++
++static int scpart_parse(struct mtd_info *master,
++			const struct mtd_partition **pparts,
++			struct mtd_part_parser_data *data)
++{
++	struct sc_part_desc *scpart_map = NULL;
++	struct mtd_partition *parts = NULL;
++	struct device_node *mtd_node;
++	struct device_node *ofpart_node;
++	struct device_node *pp;
++	const char *partname;
++	int nr_scparts;
++	int nr_parts = 0;
++	int n;
++	int res = 0;
++
++	mtd_node = mtd_get_of_node(master);
++	if (!mtd_node)
++		goto out;
++
++	ofpart_node = of_get_child_by_name(mtd_node, "partitions");
++	if (!ofpart_node)
++		goto out;
++
++	nr_scparts = scpart_find_partmap(master, &scpart_map);
++	if (nr_scparts <= 0) {
++		res = nr_scparts;
++		goto free;
++	}
++
++	for_each_child_of_node(ofpart_node, pp) {
++		u32 scpart_id;
++		struct mtd_partition *parts_tmp;
++
++		if(of_property_read_u32(pp, "scpart-id", &scpart_id))
++			continue;
++
++		for (n = 0 ; n < nr_scparts ; n++)
++			if ((scpart_map[n].part_id != ID_ALREADY_FOUND) &&
++					(scpart_id == scpart_map[n].part_id))
++				break;
++		if (n >= nr_scparts)
++			/* not match */
++			continue;
++
++		/* add the partition found in OF into MTD partition array */
++
++		/* reallocate partition array */
++		parts_tmp = parts;
++		parts = kzalloc((nr_parts + 1) * sizeof(*parts), GFP_KERNEL);
++		if (!parts) {
++			kfree(parts_tmp);
++			res = -ENOMEM;
++			goto free;
++		}
++		if (parts_tmp) {
++			memcpy(parts, parts_tmp, nr_parts * sizeof(*parts));
++			kfree(parts_tmp);
++		}
++
++		parts[nr_parts].offset = scpart_map[n].part_offs;
++		parts[nr_parts].size = scpart_map[n].part_bytes;
++		parts[nr_parts].of_node = pp;
++
++		if (!of_property_read_string(pp, "label", &partname) ||
++				!of_property_read_string(pp, "name", &partname))
++			parts[nr_parts].name = partname;
++
++		if (of_property_read_bool(pp, "read-only"))
++			parts[nr_parts].mask_flags |= MTD_WRITEABLE;
++		if (of_property_read_bool(pp, "lock"))
++			parts[nr_parts].mask_flags |= MTD_POWERUP_LOCK;
++
++		/* mark as 'done' */
++		scpart_map[n].part_id = ID_ALREADY_FOUND;
++
++		nr_parts++;
++	}
++	if (nr_parts > 0) {
++		*pparts = parts;
++		res = nr_parts;
++	} else
++		pr_info("%s: No partition in OF matches partition ID with 'SC PART MAP'.\n",
++				MOD_NAME);
++
++	of_node_put(pp);
++
++  free:
++	kfree(scpart_map);
++	if (res <= 0)
++		kfree(parts);
++
++  out:
++	return res;
++}
++
++static const struct of_device_id scpart_parser_of_match_table[] = {
++	{ .compatible = "sercomm,sc-partitions" },
++	{},
++};
++MODULE_DEVICE_TABLE(of, scpart_parser_of_match_table);
++
++static struct mtd_part_parser scpart_parser = {
++	.parse_fn = scpart_parse,
++	.name = "scpart",
++	.of_match_table = scpart_parser_of_match_table,
++};
++module_mtd_part_parser(scpart_parser);
++
++/* mtd parsers will request the module by parser name */
++MODULE_ALIAS("scpart");
++MODULE_LICENSE("GPL");
++MODULE_AUTHOR("NOGUCHI Hiroshi <drvlabo@gmail.com>");
++MODULE_DESCRIPTION("Parsing code for Sercomm/Netgear partition table");
```

**Comment by mkresin** [comment]
Sorry for the long delay. I merged you commit to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) and applied all cleanups that applied after you created the PR.

Would you please build an image from my staging tree and check if still everything works as expected.

**Comment by mkresin** [comment]
@drvlabo ping

**Comment by drvlabo** [comment]
@mkresin, thanks for merging.

I cloned sources from your staging repository and tried my own R6350.
It worked. There's no problem now.

**Comment by mkresin** [comment]
merged, thanks!
